### ISSUE TITLE
Customizable thresholds for time-based blind injection tests

### ISSUE THREAD

**Issue by FiveOFive** [zaproxy/zaproxy#8112]
### Is your feature request related to a problem? Please describe.

I've been working through false positives from the time-based blind injection tests, starting with the Remote Injection Rule (https://github.com/zaproxy/zap-extensions/blob/main/addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java) because it is using the newer linear regression algorithm.

There are some hard coded default values for the algorithm. It would be nice to be able to dynamically set these so that users can pick what works well for their scanning environment and personal tolerance for false positives.

https://github.com/zaproxy/zap-extensions/blob/main/addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/timing/TimingUtils.java#L27-L28

https://github.com/zaproxy/zap-extensions/blob/main/addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java#L220C5-L222

### Describe the solution you'd like

Ability to fully customize the time-based blind injection values.

### Describe alternatives you've considered

Changing the default time-based blind injection values. With ZAP being so widely used, it seems likely that what works well for some users may not be right for other users. So it would be difficult to get a single best setting.

Using the Threshold configuration (https://www.zaproxy.org/docs/desktop/ui/dialogs/scanpolicy/#threshold) to give High and Low hardcoded values. This could be helpful, but still doesn't allow for full customization if neither preset threshold is working well for a given application. 

### Would you like to help fix this issue?

- [X] Yes

**Comment by kingthorin** [comment]
This could be done with additional options via [Options Rule Configuration screen](https://www.zaproxy.org/docs/desktop/ui/dialogs/options/ruleconfig/). Which is probably in use for the default sleep value for many of those scan rules anyway.

**Comment by thc202** [comment]
Is it possible to have more details about the false positives? Ideally as tests for the rules.

**Comment by FiveOFive** [comment]
> Is it possible to have more details about the false positives? Ideally as tests for the rules.

Yeah let me see if I can get logs of the requests + responses and share the exact timing sequences.

**Comment by FiveOFive** [comment]
> Is it possible to have more details about the false positives? Ideally as tests for the rules.

Here are 2 examples. The tables show the series of requests made by the timing test that returned a false positive and the response times for each request.

To summarize what Remote OS Command Injection timing tests do - each test starts by sending a 1 second sleep, then 2 seconds, 3 seconds, etc... and has a [total time limit of 20 seconds](https://github.com/zaproxy/zap-extensions/blob/063cb10da70de2b4f03ed0f6e9302af43d0fe0cc/addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java#L218)

| Sleep Value (seconds) | Response Time (milliseconds) |
| --- | --- |
| 1.0 | 5377 |
| 2.0 | 6093 |
| 3.0 | 6588 |
| 4.0 | 7752 |


| Sleep Value (seconds) | Response Time (milliseconds) |
| --- | --- |
| 1.0 | 7075 |
| 2.0 | 8016 |
| 3.0 | 8488 |

The issue is that sleeping for 1, 2, or even 3 seconds is not very meaningful when scanning over the public internet or for actions with slow or variable response times. For example, if the action under test has a mean response time of 7 seconds and a standard deviation of 3 seconds, then making a few requests with small sleep times is not statistically significant.

**Proposal:** I think statistical testing is the right approach, but the default values should be updated to compare sleep times that are significantly different. For example, compare a 15 seconds sleep and a 1 second sleep instead of a 1 second sleep and 2 second sleep. This won't slow down the testing, except for positive cases, because the test can fast fail if the response time is < the requested sleep time. And it can still use the existing online simple linear regression code to calculate the correlation, but I don't think we actually need to do a step function of values, just alternative large and small.


_A couple additional notes from my investigation:_

The current code is sending unnecessary requests. It always makes requests for 1.0 seconds and 2.0 seconds and often 3.0 before failing. The [forum discussion for the statistical testing](https://groups.google.com/g/zaproxy-develop/c/KGSkNHlLtqk) includes an idea for an initial check of `actual delay > requested delay` with a larger value before proceeding to the smaller testing, but as far as I could tell this check isn't currently being done.

The sleep value in the attack parameter of the alert is incorrect. It's being set to a fixed `timeSleepSeconds` value instead of a dynamic value from the statistical testing. So the alerts always show a parameter value of 15 seconds when what actually happened was a series of requests for 1.0, 2.0, 3.0, and 4.0 seconds or 1.0, 2.0, 3.0 seconds - https://github.com/zaproxy/zap-extensions/blob/063cb10da70de2b4f03ed0f6e9302af43d0fe0cc/addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java#L652 and https://github.com/zaproxy/zap-extensions/blob/063cb10da70de2b4f03ed0f6e9302af43d0fe0cc/addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java#L603

**Comment by kingthorin** [comment]
I'm all for improving/changing these. Your proposal seems reasonable to me. You should probably also keep an eye on these:
- https://github.com/zaproxy/zap-extensions/pull/4316
- https://github.com/zaproxy/zap-extensions/pull/4997

**Comment by FiveOFive** [comment]
@kingthorin thanks for linking the related ones.

@itsdivyanshjain it looks like you and I are trying to solve the same problem.

I've opened a PR for the varied high and low delay testing (https://github.com/zaproxy/zap-extensions/pull/5053), which I think is more robust and simpler than whats in https://github.com/zaproxy/zap-extensions/pull/4997/files#diff-a85a5287173d7e22561a04af3e68c025025d9d50b7a95ea85e80041471748612 which is starting at 11 seconds and incrementing up to 15 seconds. While the starting delay is higher, that still suffers from the problem of small increments in compared delays.

**Comment by FiveOFive** [comment]
> Leaking into this issue from [zaproxy/zap-extensions#4679](https://github.com/zaproxy/zap-extensions/pull/4679). The false positive cases being addressed here are _slightly_ different, and I think both warrant explicit addressing.

Thanks for reviewing. The overload case in [zaproxy/zap-extensions#4679](https://github.com/zaproxy/zap-extensions/pull/4679) is different than what this issue describes, but I think that check is actually handled even better by the alternating high and low values.

`if (isOverload * 2.0 > lastDelay)` is sending a low sleep value and checking if it took more than twice as long as the last high value. The alternating high and low is doing that same low check, but instead of a one-time is it more than 2x check, it is adding it to the online linear regression as a data point for the total correlation. And you already have a nice optimization in the request loop to fast fail if correlation is not even close, so a low value that is more than 2 times a high value will trigger that fast fail and more cases will be handled too, beyond a hard coded 2x check.

**Comment by itsdivyanshjain** [comment]
> Leaking into this issue from [zaproxy/zap-extensions#4679](https://github.com/zaproxy/zap-extensions/pull/4679). The false positive cases being addressed here are _slightly_ different, and I think both warrant explicit addressing.
> 
> On the topic of the main improvement proposed here, I agree that an alternating approach is superior. Keeping in mind is that our goal is to **minimize FPs**, **minimize requests sent**, and **minimize time spent**:
> 
> 1. In the true negative case, the alternating approach will lead to fewer FPs, fewer total requests, and less total time spent.
> 2. In the true positive case, the alternative approach will lead to the same number of requests, and **slightly longer** time spent.
> 
> Wanting to avoid that second part lead me to avoid this approach originally, but in retrospect I do think the trade off is worth it.
> 
> I'll go ahead and provide some minor review to [PR#5053](https://github.com/zaproxy/zap-extensions/pull/5053).

Utilising alternating time do helpful in less total time spent. But, being part of product company, it's not worth to lose true positive.
Though I'm looking forward to changes if we can incorporate more robustness in detection without the tradeoff of true positive.

**Comment by thc202** [comment]
That's the plan (some of them are being changed already i.e. zaproxy/zap-extensions#4679 and zaproxy/zap-extensions#4997).

That would be appreciated!

**Comment by FiveOFive** [comment]
Remote OS Command Injection continues to look good from our side after running the new code for a week.

Here's a list of other rules that I've noticed possible timing based false positives from. I'll pick up [SQL Injection](https://www.zaproxy.org/docs/alerts/40018/) next. Didn't see a PR for it yet.

- [ ] SQL Injection | @FiveOFive looking into
- [ ] SQL Injection - Oracle - Time Based
- [ ] SQL Injection - MySQL
- [ ] SQL Injection - PostgreSQL - Time Based  | https://github.com/zaproxy/zap-extensions/pull/4997
- [ ] SQL Injection - MsSQL
- [ ] Server Side Template Injection (Blind)
- [ ] NoSQL Injection | https://github.com/zaproxy/zap-extensions/pull/4679
- [ ] Hypersonic SQL Injection
- [ ] SQLite Injection

**Comment by itsdivyanshjain** [comment]
> * Server Side Template Injection (Blind)

I have picked Mysql, SSTI as well though haven't raised PR yet. First looking forward to get merged this pending PR.

**Comment by thc202** [comment]
@FiveOFive have you started to work on the Oracle scan rule? If not I can take a look.
